File: web-client/src/services/llm/ClaudeAPIService.ts
import type { LLMMessage, LLMToken, LLMConfig } from '../../models/LLMTypes';
import type { LLMService } from '../interfaces/LLMService';

export class ClaudeAPIService implements LLMService {
  private apiKey: string;
  private abortController: AbortController | null = null;
  private readonly baseURL = 'https://api.anthropic.com/v1/messages';
  private readonly apiVersion = '2023-06-01';

  constructor(apiKey: string) {
    this.apiKey = apiKey;
  }

  abort(): void {
    this.abortController?.abort();
    this.abortController = null;
  }

  async *streamCompletion(
    messages: LLMMessage[],
    config: LLMConfig,
  ): AsyncGenerator<LLMToken, void, unknown> {
    this.abortController = new AbortController();

    // Separate system prompt from messages
    let systemPrompt: string | undefined = config.systemPrompt;
    const apiMessages: Array<{ role: string; content: string }> = [];

    for (const message of messages) {
      if (message.role === 'system') {
        systemPrompt = message.content;
      } else {
        apiMessages.push({ role: message.role, content: message.content });
      }
    }

    // Build request body
    const body: Record<string, unknown> = {
      model: config.model,
      max_tokens: config.maxTokens,
      temperature: config.temperature,
      stream: true,
      messages: apiMessages,
    };

    if (systemPrompt) {
      body.system = systemPrompt;
    }
    if (config.topP !== undefined) {
      body.top_p = config.topP;
    }
    if (config.stopSequences?.length) {
      body.stop_sequences = config.stopSequences;
    }

    const response = await fetch(this.baseURL, {
      method: 'POST',
      headers: {
        'x-api-key': this.apiKey,
        'anthropic-version': this.apiVersion,
        'content-type': 'application/json',
        'anthropic-dangerous-direct-browser-access': 'true',
      },
      body: JSON.stringify(body),
      signal: this.abortController.signal,
    });

    if (!response.ok) {
      throw this.mapResponseError(response.status);
    }

    const reader = response.body!.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let currentEventType = '';

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop()!;

        for (const line of lines) {
          if (line.startsWith('event: ')) {
            currentEventType = line.slice(7);
            continue;
          }

          if (line.startsWith('data: ')) {
            const token = this.parseSSEData(currentEventType, line.slice(6));
            if (token) {
              yield token;
              if (token.isDone) return;
            }
            currentEventType = '';
          }
        }
      }
    } finally {
      reader.releaseLock();
      this.abortController = null;
    }
  }

  private parseSSEData(_eventType: string, data: string): LLMToken | null {
    let json: Record<string, unknown>;
    try {
      json = JSON.parse(data);
    } catch {
      return null;
    }

    const type = json.type as string;

    switch (type) {
      case 'content_block_delta': {
        const delta = json.delta as Record<string, unknown> | undefined;
        const text = delta?.text as string | undefined;
        if (text) {
          return { content: text, isDone: false };
        }
        return null;
      }

      case 'message_stop':
        return { content: '', isDone: true, stopReason: 'end_turn' };

      case 'message_delta': {
        const delta = json.delta as Record<string, unknown> | undefined;
        const stopReason = delta?.stop_reason as string | undefined;
        if (stopReason) {
          return {
            content: '',
            isDone: true,
            stopReason: (stopReason === 'end_turn' || stopReason === 'max_tokens' || stopReason === 'stop_sequence')
              ? stopReason
              : 'end_turn',
          };
        }
        return null;
      }

      case 'error': {
        const error = json.error as Record<string, unknown> | undefined;
        const message = error?.message as string | undefined;
        return { content: `[Error: ${message ?? 'Unknown error'}]`, isDone: true };
      }

      default:
        return null;
    }
  }

  private mapResponseError(status: number): Error {
    switch (status) {
      case 401:
        return new Error('Invalid API key. Please check your key in Settings.');
      case 429:
        return new Error('Rate limited. Please wait a moment and try again.');
      case 400:
        return new Error('Bad request. Please check your message and try again.');
      case 404:
        return new Error('Model not found. Try selecting a different model in Settings.');
      case 529:
        return new Error('Service temporarily unavailable. Please try again later.');
      default:
        return new Error(`API error (HTTP ${status}). Please try again.`);
    }
  }
}

File: ios-client/CyranoChat/Sources/Services/LLM/ClaudeAPIService.swift
// CyranoChat - Claude API Service
// Anthropic Messages API with SSE streaming, implementing LLMService protocol

import Foundation
import Logging

/// Claude API service using the Anthropic Messages API
///
/// Supports streaming completions via Server-Sent Events (SSE).
/// Configurable model selection (Sonnet, Haiku, Opus).
public actor ClaudeAPIService: LLMService {

    // MARK: - Properties

    private let logger = Logger(label: "com.cyrano.llm.claude")
    private let apiKey: String
    private let baseURL = URL(string: "https://api.anthropic.com/v1/messages")!
    private let apiVersion = "2023-06-01"
    private let session: URLSession

    public private(set) var metrics = LLMMetrics(
        medianTTFT: 0.5,
        p99TTFT: 1.5,
        totalInputTokens: 0,
        totalOutputTokens: 0
    )

    public var costPerInputToken: Decimal {
        // Claude Sonnet: $3/M input tokens
        Decimal(string: "0.000003")!
    }

    public var costPerOutputToken: Decimal {
        // Claude Sonnet: $15/M output tokens
        Decimal(string: "0.000015")!
    }

    public var contextWindowSize: Int { 200_000 }

    // MARK: - Initialization

    public init(apiKey: String) {
        self.apiKey = apiKey
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 120
        config.timeoutIntervalForResource = 300
        self.session = URLSession(configuration: config)
        logger.info("ClaudeAPIService initialized")
    }

    // MARK: - LLMService Protocol

    public func streamCompletion(
        messages: [LLMMessage],
        config: LLMConfig
    ) async throws -> AsyncStream<LLMToken> {
        let request = try buildRequest(messages: messages, config: config)
        let startTime = Date()

        logger.info("Starting Claude stream", metadata: [
            "model": .string(config.model),
            "maxTokens": .stringConvertible(config.maxTokens)
        ])

        let (bytes, response) = try await session.bytes(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw LLMError.connectionFailed("Invalid response type")
        }

        try validateResponse(httpResponse)

        return AsyncStream { continuation in
            let task = Task {
                var isFirstToken = true
                var eventType = ""
                var dataBuffer = ""

                do {
                    for try await line in bytes.lines {
                        if Task.isCancelled { break }

                        if line.hasPrefix("event: ") {
                            eventType = String(line.dropFirst(7))
                            continue
                        }

                        if line.hasPrefix("data: ") {
                            dataBuffer = String(line.dropFirst(6))

                            if let token = parseSSEData(
                                eventType: eventType,
                                data: dataBuffer,
                                isFirstToken: &isFirstToken,
                                startTime: startTime
                            ) {
                                continuation.yield(token)
                                if token.isDone {
                                    break
                                }
                            }

                            eventType = ""
                            dataBuffer = ""
                        }
                    }
                } catch {
                    if !Task.isCancelled {
                        logger.error("Stream error: \(error.localizedDescription)")
                    }
                }

                continuation.finish()
            }

            continuation.onTermination = { @Sendable _ in
                task.cancel()
            }
        }
    }

    // MARK: - Request Building

    private func buildRequest(messages: [LLMMessage], config: LLMConfig) throws -> URLRequest {
        var request = URLRequest(url: baseURL)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue(apiVersion, forHTTPHeaderField: "anthropic-version")
        request.setValue("application/json", forHTTPHeaderField: "content-type")

        // Separate system prompt from messages
        var systemPrompt: String? = config.systemPrompt
        var apiMessages: [[String: String]] = []

        for message in messages {
            if message.role == .system {
                systemPrompt = message.content
            } else {
                apiMessages.append([
                    "role": message.role.rawValue,
                    "content": message.content
                ])
            }
        }

        // Build request body
        var body: [String: Any] = [
            "model": config.model,
            "max_tokens": config.maxTokens,
            "stream": true
        ]

        body["temperature"] = config.temperature

        if let topP = config.topP {
            body["top_p"] = topP
        }

        if let system = systemPrompt {
            body["system"] = system
        }

        if let stopSequences = config.stopSequences, !stopSequences.isEmpty {
            body["stop_sequences"] = stopSequences
        }

        body["messages"] = apiMessages

        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return request
    }

    // MARK: - Response Parsing

    private func validateResponse(_ response: HTTPURLResponse) throws {
        switch response.statusCode {
        case 200:
            return
        case 401:
            throw LLMError.authenticationFailed
        case 429:
            let retryAfter = response.value(forHTTPHeaderField: "retry-after")
                .flatMap { TimeInterval($0) }
            throw LLMError.rateLimited(retryAfter: retryAfter)
        case 400:
            throw LLMError.invalidRequest("Bad request (400)")
        case 404:
            throw LLMError.modelNotFound("Model not found")
        case 529:
            throw LLMError.connectionFailed("API overloaded (529)")
        default:
            throw LLMError.connectionFailed("HTTP \(response.statusCode)")
        }
    }

    private nonisolated func parseSSEData(
        eventType: String,
        data: String,
        isFirstToken: inout Bool,
        startTime: Date
    ) -> LLMToken? {
        guard let jsonData = data.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
              let type = json["type"] as? String else {
            return nil
        }

        switch type {
        case "content_block_delta":
            guard let delta = json["delta"] as? [String: Any],
                  let text = delta["text"] as? String else {
                return nil
            }

            var ttft: TimeInterval? = nil
            if isFirstToken {
                ttft = Date().timeIntervalSince(startTime)
                isFirstToken = false
            }
            _ = ttft // TTFT tracking for future metrics

            return LLMToken(content: text, isDone: false)

        case "message_stop":
            return LLMToken(content: "", isDone: true, stopReason: .endTurn)

        case "message_delta":
            if let delta = json["delta"] as? [String: Any],
               let stopReason = delta["stop_reason"] as? String {
                let reason = StopReason(rawValue: stopReason) ?? .endTurn
                return LLMToken(content: "", isDone: true, stopReason: reason)
            }
            return nil

        case "error":
            if let error = json["error"] as? [String: Any],
               let message = error["message"] as? String {
                // Return as a token so the UI can display it
                return LLMToken(content: "[Error: \(message)]", isDone: true)
            }
            return nil

        default:
            return nil
        }
    }

    // MARK: - Token Estimation

    public func estimateTokenCount(_ text: String) -> Int {
        // Claude uses ~3.5 chars per token for English
        max(1, text.count * 10 / 35)
    }
}
